File: agent/lib/docker.ts
import {runCommand} from "./run-command.js";

const REGISTRY = process.env.DOCKER_REGISTRY?.trim() || "";

export function useLocalOnly(): boolean {
  return REGISTRY === "";
}

export function getImageTag(serviceName: string, commitOrTag: string): string {
  const safe = sanitizeForDocker(serviceName);
  if (REGISTRY === "") return `${safe}:${commitOrTag}`;
  return `${REGISTRY}/${safe}:${commitOrTag}`;
}

export function sanitizeForDocker(name: string): string {
  return name
    .toLowerCase()
    .replace(/[^a-z0-9-]/g, "-")
    .replace(/-+/g, "-")
    .replace(/^-|-$/g, "") || "app";
}

export function detectExposedPort(imageTag: string): number | null {
  const result = runCommand(
    `docker image inspect "${imageTag}" --format '{{json .ContainerConfig.ExposedPorts}}'`,
  );
  if (!result.success) return null;
  const trimmed = (result.stdout ?? "").trim();
  if (!trimmed || trimmed === "null" || trimmed === "<no value>") return null;
  try {
    const obj = JSON.parse(trimmed);
    const keys = Object.keys(obj);
    if (keys.length === 0) return null;
    const port = parseInt(keys[0].split("/")[0], 10);
    return isNaN(port) ? null : port;
  } catch {
    return null;
  }
}

File: agent-go/internal/docker/docker.go
package docker

import (
	"encoding/json"
	"os/exec"
	"regexp"
	"strconv"
	"strings"
)

var (
	nonAlphaRe = regexp.MustCompile(`[^a-z0-9-]+`)
	dashRe     = regexp.MustCompile(`-+`)
)

// GetEnv returns the env getter (set from main with os.Getenv).
var GetEnv = func(key string) string { return "" }

// UseLocalOnly returns true when no registry is configured (skip push).
func UseLocalOnly() bool {
	return strings.TrimSpace(GetEnv("DOCKER_REGISTRY")) == ""
}

// GetImageTag returns the full image tag (registry/name:tag or name:tag).
func GetImageTag(serviceName, commitOrTag string) string {
	safe := SanitizeForDocker(serviceName)
	r := strings.TrimSpace(GetEnv("DOCKER_REGISTRY"))
	if r == "" {
		return safe + ":" + commitOrTag
	}
	return r + "/" + safe + ":" + commitOrTag
}

// DetectExposedPort checks the image for EXPOSE directives and returns the first port.
func DetectExposedPort(imageTag string) (int, bool) {
	cmd := exec.Command("docker", "image", "inspect", imageTag,
		"--format", "{{json .ContainerConfig.ExposedPorts}}")
	out, err := cmd.Output()
	if err != nil {
		return 0, false
	}
	trimmed := strings.TrimSpace(string(out))
	if trimmed == "" || trimmed == "null" || trimmed == "<no value>" {
		return 0, false
	}
	var ports map[string]interface{}
	if err := json.Unmarshal([]byte(trimmed), &ports); err != nil {
		return 0, false
	}
	for key := range ports {
		parts := strings.SplitN(key, "/", 2)
		p, err := strconv.Atoi(parts[0])
		if err == nil {
			return p, true
		}
	}
	return 0, false
}

// SanitizeForDocker normalizes a name for Docker (lowercase, alphanumeric and hyphens).
func SanitizeForDocker(name string) string {
	s := strings.ToLower(name)
	s = nonAlphaRe.ReplaceAllString(s, "-")
	s = dashRe.ReplaceAllString(s, "-")
	s = strings.Trim(s, "-")
	if s == "" {
		return "app"
	}
	return s
}
